### PR TITLE
sys_paths: use XDG_CONFIG_HOME if present, new API

### LINKED ISSUE #2020: Quadrature Demod: missing FM input end-to-end QA test
So, we need a unit test with FM modulated signals getting decoded correctly in a zero-noise scenario.

extend gr-analog/python/analog/qa_quadrature_demod.py:

1. add a new `def test_fm_end_to_end(self):`
2. in that, use the `analog.frequency_modulator_fc` block with, 
3. implement 
```text 
signal source (cos 1000 Hz) +-> vector_sink "A"
                            |
                            \-> analog.frequency_modulator_fc -> analog.quadrature_demod_cf -> vector_sink "B"
```
4. verify that, first samples aside, A.data() is equal to B.data() after running

### LINKED ISSUE #3631: Configuration files should be stored following the XDG Base Directory Specification
For some reason gnuradio creates a folder `~/.gnuradio` and uses it for configuration. That's not in-line with [XDG Base Directory Specification](https://wiki.debian.org/XDGBaseDirectorySpecification) and clutters people's home directories.

gnuradio should follow XDG base directory specification and use `~/.config/gnuradio` for configuration files

### PR DESCRIPTION
**Pull request by marcusmueller**
## Description
- Introduced `gr::paths::` namespace to keep the different path getters
  together
  - new `cache()` getter, because FFTW wisdom should not clutter the
    home directory, but is also not portable state nor portable config
    (compare
    [`XDG_CACHE_HOME`](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html))
- new path getters return `std::filesystem::path` instead of C string
- deprecated old getters, which now just call the new getters underneath
- Python wrappers
- added a unit test for the paths (was untested so far)
- added prefs path unit test for checking reaction to changing
  environment variables (prefs was completely untested)
- Fixed `vmcircbuf_prefs`, fftw wisdom that assumed existence of paths
- Replace deprecated API from in-tree usage
- As some configuration files move to subdirectories of directories not
  guaranteed to exist, create directories as needed

In this new configuration, `gr::paths::userconf()` (and the legacy,
`gr::userconf_path()`) tries to find the user configuration in the
following order:

1. `$GR_PREFS_PATH` if that environment variable is set
2. `$XDG_CONFIG_HOME/gnuradio`; if `$XDG_CONFIG_HOME` is set, and
   `$XDG_CONFIG_HOME/gnuradio` exists and is a directory
3. `appdata()/.config/gnuradio` (`appdata ~= $HOME`) if that exists and
   is a directory
4. `appdata()/.gnuradio` if that exists and is a directory
5. `$XDG_CONFIG_HOME/gnuradio` if `$XDG_CONFIG_HOME` is set
6. `appdata()/.config/gnuradio`

So, in other words: unless explicitly set to something else, try the new
XDG_CONFIG_HOME path, if that doesn't exist, try the old path, and if
neither exists, use the new one.
## Related Issue

Supersedes #7073

(but still uses mboersch code, so copyright notices include mboersch)

Closes #3631

Looks at #1010, but makes only minor progress

## Which blocks/areas does this affect?

- `gr::userconf_path()`: deprecated
- `gr::tmp_path()`: deprecated
- `gr::appdata_path()`: deprecated
- `gr::fft::fft`: stores wisdom no longer in `~/.gr_fftw_wisdom` but in
  `~/.cache/gnuradio/fftw_wisdom`
- `vmcircbuf_prefs`: instead of hard-coding `appdata()/.gnuradio/prefs/key`,
  uses `userconf()/prefs/key` (which is a bugfix, actually)

## Testing Done

- wrote a sys path test
- wrote prefs tests (long overdue)

## Checklist

- [x] I have read the [CONTRIBUTING document](https://github.com/gnuradio/gnuradio/blob/main/CONTRIBUTING.md).
- [x] I have squashed my commits to have one significant change per commit. 
- [x] I [have signed my commits before making this PR](https://github.com/gnuradio/gnuradio/blob/main/CONTRIBUTING.md#dco-signed)
- [x] My code follows the code style of this project. See [GREP1.md](https://github.com/gnuradio/greps/blob/main/grep-0001-coding-guidelines.md).
- [x] I have updated [the documentation](https://wiki.gnuradio.org/index.php/Main_Page#Documentation) where necessary.
- [x] I have added tests to cover my changes, and all previous tests pass.

## TODO

After merge of this: 

- Document on wiki (@duggabe will like that)

### CODE REVIEW

**Comment by marcusmueller** [comment]
@willcode  no API change, only ABI-irrelevant annotations, and additions: designed for backportability.

**Comment by marcusmueller** [comment]
really hoping this works on other platforms as well!

**Comment by marcusmueller** [comment]
oha.

so, situation:

- because the parent directory doesn't inherently exist, need to create ~/.config/gnuradio recursively
- Windows is fun: https://en.cppreference.com/w/cpp/filesystem/path/generic_string ; need to `path::string().c_str()` to get the path in the legacy functions. Oh well.

**Comment by marcusmueller** [comment]
TODO: check whether CI runs all fail during testing or whether something strange is going on; some fail surprisingly fast.

File: gnuradio-runtime/include/gnuradio/sys_paths.h [hunk 1]
```diff
@@ -10,18 +11,51 @@
...
+*/
+} /* namespace  paths */
 
 //! directory to create temporary files
-GR_RUNTIME_API const char* tmp_path();
+[[deprecated("use gr::paths::tmp")]] GR_RUNTIME_API const char* tmp_path();
 
 //! directory to store application data
-GR_RUNTIME_API const char* appdata_path();
+[[deprecated("use gr::paths::tmp")]] GR_RUNTIME_API const char* appdata_path();
```
Comment (willcode) [inline gnuradio-runtime/include/gnuradio/sys_paths.h:55]: Deprecation note on the last 2 - cut/paste.

File: gnuradio-runtime/include/gnuradio/sys_paths.h [hunk 2]
```diff
@@ -10,18 +11,51 @@
...
 #include <gnuradio/api.h>
+#include <filesystem>
 
 namespace gr {
+namespace paths {
+/*! \brief directory to create temporary files.
+ *
+ * On UNIX-oid systems, typically /tmp.
+ */
+GR_RUNTIME_API std::filesystem::path tmp();
```
Comment (willcode) [inline gnuradio-runtime/include/gnuradio/sys_paths.h:24]: Does the name change provide enough value to bother people with deprecation warnings?

Comment (marcusmueller) [reply gnuradio-runtime/include/gnuradio/sys_paths.h:24]: yes. The main point is that the return value changes, from a manually (actually, leaking or use after free or concurrently modified in the current API) `const char*` to a `std::filesystem::path`.

**Comment by willcode** [comment]
F39 CI is failing possibly due to the gr-fft portion of this fix. Hold backport until we figure it out.

**Comment by willcode** [comment]
Backport: add #7181
